src/OrbitBase/include/OrbitBase/WhenAny.h

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_BASE_WHEN_ANY_H_
+#define ORBIT_BASE_WHEN_ANY_H_
+
+#include <absl/base/thread_annotations.h>
+#include <absl/synchronization/mutex.h>
+
+#include <variant>
+
+#include "OrbitBase/Future.h"
+#include "OrbitBase/FutureHelpers.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/Promise.h"
+
+namespace orbit_base_internal {
+
+// A small helper type trait that maps `void` to `std::monostate` and every other `T` to itself.
+template <typename T>
+struct VoidToMonostate {
+  using type = T;
+};
+
+template <>
+struct VoidToMonostate<void> {
+  using type = std::monostate;
+};
+
+template <typename T>
+using VoidToMonostate_t = typename VoidToMonostate<T>::type;
+
+template <typename... Ts>
+class SharedStateWhenAny {
+  orbit_base::Promise<std::variant<Ts...>> promise ABSL_GUARDED_BY(mutex);
+  mutable absl::Mutex mutex;
+
+ public:
```

Comment:
We usually prefer the `public:` part at the top.

Reply:
Yep. Done.